File: Api/Transfer/Project.cs
using System;
using System.Collections.Immutable;
using Kafe.Data;

namespace Kafe.Api.Transfer;

/// <summary>
/// A DTO of a project. To be used when listing projects.
/// </summary>
/// <param name="GlobalPermissions">Permissions that apply to all users, even the anonymous ones.</param>
/// <param name="UserPermissions">
/// Permissions that apply to the currently logged in user. Includes the global permissions.
/// </param>
public record ProjectListDto(
    Hrib Id,
    Hrib ProjectGroupId,
    LocalizedString Name,
    LocalizedString? Description,
    ImmutableArray<Permission> GlobalPermissions,
    ImmutableArray<Permission> UserPermissions,
    DateTimeOffset ReleasedOn,
    ReviewKind LatestReviewKind,
    bool IsLocked,
    Hrib? OwnerId
    // TODO: Thumbnail
);

public record ProjectDetailDto(
    Hrib Id,
    Hrib ProjectGroupId,
    LocalizedString? ProjectGroupName,
    ProjectValidationSettings ValidationSettings,
    LocalizedString? Genre,
    LocalizedString Name,
    LocalizedString? Description,
    string? AiUsageDeclaration,
    string? HearAboutUs,
    ImmutableArray<Permission> GlobalPermissions,
    ImmutableArray<Permission> UserPermissions,
    DateTimeOffset ReleasedOn,
    ImmutableArray<ProjectAuthorDto> Crew,
    ImmutableArray<ProjectAuthorDto> Cast,
    Hrib? OwnerId,
    ImmutableArray<ProjectArtifactDto> Artifacts,
    ImmutableArray<ProjectReviewDto> Reviews,
    ProjectBlueprintDto Blueprint,
    bool IsLocked
);

public record ProjectAuthorDto(
    Hrib Id,
    string Name,
    ImmutableArray<string> Roles
);

public record ProjectArtifactDto(
    Hrib Id,
    LocalizedString Name,
    DateTimeOffset AddedOn,
    string? BlueprintSlot,
    ImmutableArray<ShardListDto> Shards
);

public record ProjectReviewDto(
    Hrib? ReviewerId,
    ReviewKind Kind,
    string ReviewerRole,
    LocalizedString? Comment,
    DateTimeOffset AddedOn
);

public record ProjectReviewCreationDto(
    Hrib ProjectId,
    ReviewKind Kind,
    string ReviewerRole,
    LocalizedString? Comment
);

public record ProjectBlueprintDto(
    ImmutableArray<string> RequiredReviewers,
    ImmutableDictionary<string, ProjectArtifactBlueprintDto> ArtifactBlueprints
);

public record ProjectArtifactBlueprintDto(
    LocalizedString Name,
    LocalizedString? Description,
    ArgumentArity Arity,
    ImmutableDictionary<ShardKind, ProjectArtifactShardBlueprintDto> ShardBlueprints
);

public record ProjectArtifactShardBlueprintDto(
    LocalizedString Name,
    LocalizedString? Description,
    ArgumentArity Arity
);

public record ProjectCreationDto(
    Hrib ProjectGroupId,
    LocalizedString Name,
    LocalizedString? Description,
    LocalizedString? Genre,
    string? AiUsageDeclaration,
    string? HearAboutUs,
    ImmutableArray<ProjectCreationAuthorDto> Crew,
    ImmutableArray<ProjectCreationAuthorDto> Cast,
    bool IsLocked
);

public record ProjectEditDto(
    Hrib Id,
    LocalizedString? Name,
    LocalizedString? Description,
    LocalizedString? Genre,
    string? AiUsageDeclaration,
    string? HearAboutUs,
    ImmutableArray<ProjectCreationAuthorDto>? Crew,
    ImmutableArray<ProjectCreationAuthorDto>? Cast,
    ImmutableArray<ProjectArtifactAdditionDto>? Artifacts,
    bool? IsLocked
);

public record ProjectCreationAuthorDto(
    Hrib Id,
    ImmutableArray<string> Roles
);

public record ProjectValidationDto(
    Hrib ProjectId,
    DateTimeOffset ValidatedOn,
    ImmutableArray<ProjectDiagnosticDto> Diagnostics
);

public record ProjectDiagnosticDto(
    DiagnosticKind Kind,
    LocalizedString Message,
    string ValidationStage
);

public record ProjectArtifactAdditionDto(
    Hrib Id,
    string? BlueprintSlot
);

File: Web/src/data/Project.ts
import { DateTime } from 'luxon';
import { components } from '../schemas/api';
import { localizedString } from '../schemas/generic';
import { getPrefered } from '../utils/preferedLanguage';
import { AbstractType } from './AbstractType';
import { Artifact } from './Artifact';
import { localizedMapper } from './serialize/localizedMapper';
import { rolesMapper } from './serialize/rolesMapper';
import { Serializer } from './serialize/Serializer';
import { TaggedOption } from '@/utils/managers/common';
import genreManager from '@/utils/managers/genreManager';

export class Project extends AbstractType {
    // API object
    public projectGroupId!: string;
    public projectGroupName!: localizedString;
    public validationSettings?: components['schemas']['ProjectValidationSettings'];
    public name?: localizedString;
    public genre?: localizedString;
    public genreTags?: TaggedOption[];
    public description?: localizedString;
    // public visibility!: components['schemas']['Visibility'];
    public releasedOn!: Date | null;

    // Authors
    public crew!: components['schemas']['ProjectAuthorDto'][];
    public cast!: components['schemas']['ProjectAuthorDto'][];
    public ownerId?: string;

    public artifacts!: Artifact[];

    public reviews!: components['schemas']['ProjectReviewDto'][];
    public blueprint!: components['schemas']['ProjectBlueprintDto'];
    public globalPermissions!: components['schemas']['Permission'];
    public userPermissions!: components['schemas']['Permission'];
    public isLocked!: boolean;
    public latestReviewKind?: components['schemas']['ReviewKind'];

    public constructor(struct: components['schemas']['ProjectListDto'] | components['schemas']['ProjectDetailDto']) {
        super();
        Object.assign(this, struct);
        this.releasedOn = new Date(struct.releasedOn);
        if (this.releasedOn.getTime() < 0) this.releasedOn = null;
        if (this.artifacts) {
            this.artifacts = this.artifacts.map((artifact: any) => new Artifact(artifact));
        }

        this.genreTags = genreManager.getValue(this.genre);

        // Temporary
        // this.blueprint.artifactBlueprints = (
        //     this.blueprint.artifactBlueprints as any as components['schemas']['ProjectArtifactBlueprintDto'][]
        // ).reduce(
        //     (prev, curr) => ({
        //         ...prev,
        //         [(curr as any).slotName]: (
        //             curr.shardBlueprints as any as components['schemas']['ProjectArtifactShardBlueprintDto'][]
        //         ).reduce((pr, c) => ({ ...pr, [(c as any).kind]: c }), {}),
        //     }),
        //     {},
        // );
    }

    public getName() {
        return getPrefered(this.name);
    }

    public getDescription() {
        return getPrefered(this.description);
    }

    public getGroupName() {
        return getPrefered(this.projectGroupName);
    }

    public getGenre() {
        return getPrefered(this.genre);
    }

    public getLastUpdate(): Date | null {
        const lastUpdate = this.artifacts
            .filter((artifact) => artifact.addedOn !== null)
            .sort((a, b) => b.addedOn!.getTime() - a.addedOn!.getTime())[0]?.addedOn;
        return lastUpdate ? lastUpdate : null;
    }

    serialize(update: boolean = false): components['schemas']['ProjectCreationDto'] {
        return new Serializer(this, update)
            .addConditionaly(!update, 'projectGroupId')
            .add('name', localizedMapper)
            .add('genre', localizedMapper)
            .add('description', localizedMapper)
            .add('artifacts', (artifacts: Artifact[]) =>
                artifacts ? artifacts.map((artifact) => artifact.serialize(false)) : undefined,
            )
            .add('releasedOn', (date: Date | null) => date ? DateTime.fromJSDate(date).toISO() : null)
            .add('crew', rolesMapper)
            .add('cast', rolesMapper)
            .build();
    }
}
